Arm/Memory/Separate.lean

```diff
@@ -115,6 +116,245 @@ example : mem_subset (BitVec.ofNat 64 (2^64 - 1)) 0#64 0#64 (BitVec.ofNat 64 (2^
 def mem_legal (a1 a2 : BitVec 64) : Bool :=
   a1 <= a2
 
+/-
+Given two legal memory regions `[a1, a2]` and `[b1, b2]` that are separated, 
+it must be the case either:
+  - the first one ends before the second one starts (`a2 < b1`),
+  - or, the first one starts after the second one ends `(a1 > b2)`,
+-/
+theorem lt_or_gt_of_mem_separate_of_mem_legal_of_mem_legal (h : mem_separate a1 a2 b1 b2)
+    (ha : mem_legal a1 a2) (hb : mem_legal b1 b2) :
+    a2 < b1 ∨ a1 > b2 := by
+  unfold mem_separate mem_overlap at h
+  obtain ⟨⟨⟨h₁, h₂⟩, h₃⟩, h₄⟩ := by simpa? using h
+  simp [mem_legal] at ha hb
+  rw [BitVec.le_def] at ha hb
+  rw [BitVec.le_def] at h₁ h₂ h₃ h₄
+  rw [BitVec.lt_def, BitVec.gt_def]
+  by_cases h₅ : a2.toNat < b1.toNat
+  · simp [h₅]
+  · by_cases h₆ : a1.toNat > b2.toNat
+    · simp [h₆]
+    · exfalso
+      rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le ha] at h₁ h₂
+      rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le hb] at h₃ h₄
+      have h₅' : b1.toNat ≤ a2.toNat := by omega
+      have h₆' : a1.toNat ≤ b2.toNat := by omega
+      rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le h₅'] at h₄
+      rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le h₆'] at h₂
+      omega
+
+@[simp]
+theorem BitVec.neg_neg (x : BitVec w₁) : - (- x) = x := by 
```

Comment:
I am currently upstreaming this. A review comment is much appreciated:

https://github.com/leanprover/lean4/pull/4977